### PR TITLE
Shopping Cart Data Structure

### PR DESCRIPTION
**Pull request by thinkocapo**
## Overview
You can now click on product items and they get added into the shopping cart data structure (see the console logs in XCode). The Shopping Cart UI itself is still not yet built, but that should be easy. It's just another table.

#### Features
- **Singleton class ShoppingCart.swift** as the datamodel/database for shopping cart data, implemented, and its data (static variables) can be updated and accessed from any controller.
- **TableView** click handler for selecting a row (i.e. adding item to shopping cart). followed [this tutorial](https://www.simplifiedios.net/ios-tableview-tutorial-handling-clicks/) and updated method names according to IDE's warnings+suggestions.

Data Structure for shopping cart, borrowed from the application-monitoring-javascript demo
image

Did not add methods for Removing items from Shopping Cart yet. You can restart the app for now. We also don't capture item deletions as button clicks in breadcrumbs yet (our tests don't do that).

#### Enhancements
- Product Entity model updated with remaining attributes
- utility function for resetting database
- products only write to database if database is empty. but http request + db query still run every time (for the spans in the transaction)

## Testing
[EmpowerPlantViewController transaction](https://sentry.io/organizations/testorg-az/discover/will-cocoa:2a953cba7c14476ba556ce96a890f2f5/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&project=6249899&query=&sort=-timestamp&statsPeriod=1h&yAxis=count%28%29)
[CartViewController transaction](https://sentry.io/organizations/testorg-az/discover/will-cocoa:eaef78c3945348c8bd7f0849790de670/?field=title&field=event.type&field=project&field=user.display&field=timestamp&name=All+Events&project=6249899&query=&sort=-timestamp&statsPeriod=1h&yAxis=count%28%29)

Empower Plant home screen
image

Shopping Cart
image

log of Total being computed, and it saves to the ShoppingCart.swift singleton
image

## Next
Add 'TOTAL' price to the Shopping Cart UI screen.

A checkout form (optional, some mobile demo's don't have it) for name/address/phone.

Add a Purchase button.

For checkout (Purchase), will need put these on a 'cart' property on req body, may need a JSON object for that, which has Id's again

        /*
         // how it works in application-monitoring-javascript
         body: JSON.stringify({
             cart: cart,
             form: this.state
         })
         */

can try building JSON like this:
```
// EXAMPLE
let jsonObject: [String: Any] = [
            "type_id": 1,
            "model_id": 1,
            "transfer": [
                "startDate": "10/04/2015 12:45",
                "endDate": "10/04/2015 16:00"
            ],
            "custom": "savedData"
        ]
```
## Technical Notes
Data persistence options for Shopping Cart...

ModelController - Too much boilerplate...https://code.tutsplus.com/tutorials/the-right-way-to-share-state-between-swift-view-controllers--cms-28474

Data Container - Too much boilerplate...

Singleton Pattern - Easy, https://betterprogramming.pub/5-ways-to-pass-data-between-view-controllers-18acb467f5ec class Settings

File I/O - Gets us a span for the transaction but is overkill for a ShoppingCart. We can add a File I/O example elsewhere.

CoreData - Easy (already using)

Segue - Going back and forth between controllers, could present problems with persistence. Sounds like it's intended more for one-way data flow https://levelup.gitconnected.com/swift-xcode-sharing-data-between-view-controllers-8d270e99ca1e

Also....
```
// These failed
var quantities = {}
class quantities: NSObject {}
var quantities: NSObject = {}
struct quantities {}
var quantities = {
"1": 0;
"2": 0;
}
struct quantities {
var one = 0
var two = 0
}
https://stackoverflow.com/questions/31145990/dynamically-create-objects-and-set-attributes-in-swift. 
https://stackoverflow.com/questions/38121957/singleton-and-class-properties-in-swift
```
and
```
self.instance.quantities.setValue(0, forKey: "key")
self.instance.quantities['prop1'] = 0
self.instance.quantities[product.productId] = self.instance.quantities[product.productId] || 0
self.instance.quantities["prop1"] = 0 // struct
```

https://stackoverflow.com/questions/26667380/in-swift-for-anyobject-how-do-i-setvalue-then-call-valueforkey

### CODE REVIEW

File: EmpowerPlant/ShoppingCart.swift [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        
+        // These are the values as used in the backend database in CloudSQL
+        /*
+         Plant Mood 3
+         Botana Voice 4
+         Plant Stroller 5
+         Plant Nodes 6
+        */
+        switch id {
+        case 3:
```
Comment (cstavitsky) [inline EmpowerPlant/ShoppingCart.swift:41]: (note: this is not a blocker for approving this PR, just a thought)

The repetition of these kinds of sections several times seems not-ideal:

```
self.instance.quantities.plantMood += 1
let price = Int(product.price!)
self.instance.total = self.instance.total + price!
print("> quantities plantMood", self.instance.quantities.plantMood)
break
```

I was originally thinking we could clean this up and not repeat code. Maybe dynamically call `self.instance.quantities.<the product name>`, i.e. call a function based on a string. I didn't know the 'Swift way' to do this, but when I googled it seemed like it's actively discouraged from doing that. They offer another idea, but I'm not sure that really buys us much 🤷  https://www.hackingwithswift.com/forums/swift/calling-object-methods-dynamically/1983

All this is to say that I think this section could possibly be cleaned up, but I don't know the right way of doing it without doing a bunch of additional research (which is not worth it because this is straightforward/contained enough that it doesn't actually create a problem), so I have nothing valuable to add and am fine to move forward with what's here. 😄

File: EmpowerPlant/EmpowerPlantViewController.swift [hunk 2]
```diff
@@ -31,36 +33,125 @@ class EmpowerPlantViewController: UIViewController, UITableViewDelegate, UITable
...
+        
+        self.navigationItem.leftBarButtonItem = UIBarButtonItem(
+            title: "List App",
+            style: .plain,
+            target: self,
+            action: #selector(goToListApp) // clearDb
+        )
+    }
+    
+    // Writes to CoreData database
```
Comment (cstavitsky) [inline EmpowerPlant/EmpowerPlantViewController.swift:103]: @thinkocapo can you remind me why we're creating products in the iOS database instead of fetching products from one of the existing backends that have their own databases? I was assuming mobile would be treated like a frontend/client application in the same way our react app is, so wanted to know if there is a reason for using the DB in the iOS app itself?

**Review by cstavitsky — APPROVED** [review]

File: EmpowerPlant/EmpowerPlantViewController.swift (hunk 2 above)
Comment (thinkocapo) [reply EmpowerPlant/EmpowerPlantViewController.swift:103]: We're doing it because the sentry-cocoa sdk captures the db queries as spans for the transactions.

It kind of emulates caching, saving state/stuff locally rather than having to fetch. Not fully realistic but it's better than nothing or writing dummy data queries. 

If user navigates away, but then comes back, and has since lost wifi, then no http call is needed, because the data was saved locally.

@cstavitsky

File: EmpowerPlant/ShoppingCart.swift (hunk 1 above)
Comment (thinkocapo) [reply EmpowerPlant/ShoppingCart.swift:41]: @cstavitsky 
I can try `self.instance.quantities.<the product name>` but the problem is that's a registered property at compile-time. Swift doesn't allow things like self.instance.quantities["propertyName"]. I know there's a getter for this and as long as it existed at compile-time, it may access it. But that's only for 'getting' and not 'setting'. I tried for an hour to build those ["propertyNames"]'s on the fly, so it'd make 4 if it's 4 products or 1,000 if it's 1,000 products but couldn't get it to work. Here's an [explanation about the problem in swift](https://stackoverflow.com/questions/26667380/in-swift-for-anyobject-how-do-i-setvalue-then-call-valueforkey). I could try changing the data structure altogether, but then I'd have to later on transform to match the backend's data structure. Better to keep it the same as what we have in JS, python, java, etc.

The real data structure (coming from the Empower Plant in javascript) has numbers as keys, for the quantity, and that's not allowed in swift, like it is in Javascript. So this is why I ended up with this weird data structure `self.instance.quantities.<the product name>` in ShoppingCart.swift, and the two switch statements you saw, which are accomplishing different things. I had to convert that dictionary to an array for the shopping cart UI, because the Table works off of an array (and not a dictionary, which is what Empower Plant in javascript uses).

I looked at your link and it says "Swift doesn't really do dynamism like that." yeah it's a lot of code to accomplish something simple. It was hard to follow. Switch statements and declaring the products as properties works as it's only 4 products and not 10,000 products. No plans for +4 products any time soon (creative team created those).

Yeah I can put a helper function in there for that switch statement.

Happy to present the weirdness about it on a call. It's definitely not ideal but works for now.

Comment (thinkocapo) [reply EmpowerPlant/ShoppingCart.swift:41]: I'm trying to make
```
        case 6:
            updateQuantitiesAndTotal(self.instance.quantities.plantNodes, product)
            break
```

with
```
    static func updateQuantitiesAndTotal(productQuantity: <Type>, product: product) {
       // this doesn't work, and I can't find a way to do it with Generics in Swift yet
       self.instance.quantities[<Type>] += 1
            let price = Int(product.price!)
            self.instance.total = self.instance.total + price!
        print("> updateQuantitiesAndTotal")
    }
```
Also tried
```
    static func updateQuantitiesAndTotal(productName: String, product: Product) {
       // this doesn't compile, hence I thought about using Generic Types
        self.instance.quantities["productName"] += 1
            let price = Int(product.price!)
            self.instance.total = self.instance.total + price!
        print("> updateQuantitiesAndTotal")
    }
```


Ok so found a way to separate out the updateTotal at least. That's what you'll see in the new code.


I tried to do updateQuantitiesAndTotal but ran into the same problems as before, and the code started getting hairy, losing readability, and needing yet more switch statements and helpers
image

File: EmpowerPlant/EmpowerPlantViewController.swift (hunk 2 above)
Comment (thinkocapo) [reply EmpowerPlant/EmpowerPlantViewController.swift:103]: But it also fetches the products from the backend, so we get a span for that. If the database is empty, then it writes them to the database. But need db query + http request every single time.

The point is to get spans. The business logic and what's going on is opaque to the user/audience.
